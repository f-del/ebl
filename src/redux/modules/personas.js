/***
 *    ██████╗ ███████╗██████╗ ██╗   ██╗ ██████╗███████╗██████╗ ███████╗
 *    ██╔══██╗██╔════╝██╔══██╗██║   ██║██╔════╝██╔════╝██╔══██╗██╔════╝
 *    ██████╔╝█████╗  ██║  ██║██║   ██║██║     █████╗  ██████╔╝███████╗
 *    ██╔══██╗██╔══╝  ██║  ██║██║   ██║██║     ██╔══╝  ██╔══██╗╚════██║
 *    ██║  ██║███████╗██████╔╝╚██████╔╝╚██████╗███████╗██║  ██║███████║
 *    ╚═╝  ╚═╝╚══════╝╚═════╝  ╚═════╝  ╚═════╝╚══════╝╚═╝  ╚═╝╚══════╝
 *
 */

const initialState = {
  list: {
    e6wTJIBsAw5puwwrNEsR: {
      Name: "Feature Team",
      Details: "Meta persona, representing all collaborators working with Paul",
      Needs: [
        "To understand WHAT the Product Owner wants to be able to design collaborativly the HOW",
        "To manage easly the list of things to do to finish the most quickly the User Story"
      ],
      Picture: ""
    },
    uOCF2GLFiJl6fvxdlq4b: {
      Details: "Product Owner for a new product, work in a startup",
      Name: "Paul Deligny",
      Needs: [
        "To have a clear understanding about the product vision & objectives",
        "To write real User Stories, ie have a Customer Driven thinking",
        "To share at the right time and with the rigth level of information the User Stories with his team",
        "To have a clear vision on the job he'll have to do AND the job already or currently done"
      ],
      Picture: ""
    }
  }
};

export default function(state = initialState, action) {
  return state;
}

/***
 *    ███████╗███████╗██╗     ███████╗ ██████╗████████╗ ██████╗ ██████╗ ███████╗
 *    ██╔════╝██╔════╝██║     ██╔════╝██╔════╝╚══██╔══╝██╔═══██╗██╔══██╗██╔════╝
 *    ███████╗█████╗  ██║     █████╗  ██║        ██║   ██║   ██║██████╔╝███████╗
 *    ╚════██║██╔══╝  ██║     ██╔══╝  ██║        ██║   ██║   ██║██╔══██╗╚════██║
 *    ███████║███████╗███████╗███████╗╚██████╗   ██║   ╚██████╔╝██║  ██║███████║
 *    ╚══════╝╚══════╝╚══════╝╚══════╝ ╚═════╝   ╚═╝    ╚═════╝ ╚═╝  ╚═╝╚══════╝
 *
 */
export function getAllPersonas(state) {
  return mapToArrayPersona(getPersonasList(state)) || [];
}

function getPersonasList(state) {
  return getPersonas(validateState(state)).list || {};
}

/* Functionnal helper */
function validateState(state) {
  if (state === undefined) throw new Error("State arg is mandatory");
  return state;
}
function getPersonas(state) {
  return state.personas || {};
}

function mapToArrayPersona(list) {
  return Array.from(Object.keys(list), key => ({
    ...getPersonaId(key),
    ...list[key]
  }));
}
function getPersonaId(key) {
  return {
    Id: key
  };
}
