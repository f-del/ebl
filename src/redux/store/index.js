import cards from "../modules/cards";
import { combineReducers } from "redux";

const rootReducer = combineReducers({ cards });
export default rootReducer;
