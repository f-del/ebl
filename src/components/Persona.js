import React from "react";
import PropTypes from "prop-types";
import CreateHypothesisCard from "../containers/CreateHypothesisCard";
import CardListByStatus from "../containers/CardsListByStatus";
import { cardStatus, cardType } from "../redux/modules/cards";
import CardsBoard from "../containers/CardsBoard";

function Persona({ persona, addStory, selectedNeeds }) {
  const filtredNeeds = () => {
    if (selectedNeeds !== undefined)
      return persona.Needs.filter((n, i) => i === selectedNeeds);
    else return persona.Needs;
  };
  return (
    <div>
      {persona.Name}
      <ul>
        {filtredNeeds().map((p, i) => (
          <li key={i}>
            {p}
            {addStory && (
              <React.Fragment>
                <CreateHypothesisCard
                  personaId={persona.Id}
                  personaNeedsIndex={i}
                />
                <CardsBoard type={cardType.Hypothesis}>
                  <CardListByStatus
                    type={cardType.Hypothesis}
                    status={cardStatus.TODO}
                    filterBy={{ persona: persona.Id, needsIndex: i }}
                  />
                </CardsBoard>
              </React.Fragment>
            )}
          </li>
        ))}
      </ul>
    </div>
  );
}

Persona.propTypes = {
  persona: PropTypes.object.isRequired,
  addStory: PropTypes.bool.isRequired
};

export default Persona;
