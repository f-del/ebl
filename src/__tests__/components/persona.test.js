import React from "react";
import { shallow, configure, mount } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { entity_persona_created } from "../redux/personas";
import Persona from "../../components/Persona";
import { createStore, applyMiddleware } from "redux";
import { Provider } from "react-redux";
import thunk from "redux-thunk";
import reducer from "../../redux/store/index";

configure({ adapter: new Adapter() });
var store;
describe("Componant tests", () => {
  beforeEach(() => {
    jest.resetAllMocks();
    store = createStore(reducer, applyMiddleware(thunk));
  });

  test.skip("Display a persona without children", () => {
    const wrapper = shallow(
      <Persona persona={entity_persona_created} addStory={false} />
    );

    expect(wrapper.text()).toContain(entity_persona_created.Name);

    const needsList = wrapper.find("li");
    expect(needsList.length).toBe(entity_persona_created.Needs.length);
    expect(needsList.at(0).text()).toContain(entity_persona_created.Needs[0]);
    expect(wrapper.find("button").length).toBe(0);
  });

  test.skip("Display a persona with children", () => {
    const wrapper = mount(
      <Provider store={store}>
        <Persona persona={entity_persona_created} addStory={true} />
      </Provider>
    );

    expect(wrapper.text()).toContain(entity_persona_created.Name);

    const needsList = wrapper.find("li");
    expect(needsList.length).toBe(entity_persona_created.Needs.length);
    expect(needsList.at(0).text()).toContain(entity_persona_created.Needs[0]);

    expect(wrapper.find("button").length).toBe(4);

    wrapper.unmount();
  });
});
