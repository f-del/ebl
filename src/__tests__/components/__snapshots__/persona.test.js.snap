// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Componant tests Display a persona and a specific needs 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Persona
    addStory={false}
    persona={
      Object {
        "Details": "Product Owner for a new product, work in a startup",
        "Id": "uOCF2GLFiJl6fvxdlq4b",
        "Name": "Paul Deligny",
        "Needs": Array [
          "To have a clear understanding about the product vision & objectives",
          "To write real User Stories, ie have a Customer Driven thinking",
          "To share at the right time and with the rigth level of information the User Stories with his team",
          "To have a clear vision on the job he'll have to do AND the job already or currently done",
        ],
        "Picture": "",
      }
    }
    selectedNeeds={1}
  >
    <div>
      Paul Deligny
      <ul>
        <li
          key="0"
        >
          To write real User Stories, ie have a Customer Driven thinking
        </li>
      </ul>
    </div>
  </Persona>
</Provider>
`;

exports[`Componant tests Display a persona with children 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Persona
    addStory={true}
    persona={
      Object {
        "Details": "Product Owner for a new product, work in a startup",
        "Id": "uOCF2GLFiJl6fvxdlq4b",
        "Name": "Paul Deligny",
        "Needs": Array [
          "To have a clear understanding about the product vision & objectives",
          "To write real User Stories, ie have a Customer Driven thinking",
          "To share at the right time and with the rigth level of information the User Stories with his team",
          "To have a clear vision on the job he'll have to do AND the job already or currently done",
        ],
        "Picture": "",
      }
    }
  >
    <div>
      Paul Deligny
      <ul>
        <li
          key="0"
        >
          To have a clear understanding about the product vision & objectives
          <Connect(CreateCard)
            personaId="uOCF2GLFiJl6fvxdlq4b"
            personaNeedsIndex={0}
          >
            <CreateCard
              onValidate={[Function]}
              personaId="uOCF2GLFiJl6fvxdlq4b"
              personaNeedsIndex={0}
            >
              <div
                className="create-card"
              >
                <button
                  onClick={[Function]}
                >
                  Create new task
                </button>
              </div>
            </CreateCard>
          </Connect(CreateCard)>
          <Connect(CardsBoard)
            type="HYPOTHESIS"
          >
            <CardsBoard
              getCards={[Function]}
              type="HYPOTHESIS"
            >
              <div>
                <Connect(CardsList)
                  filterBy={
                    Object {
                      "needsIndex": 0,
                      "persona": "uOCF2GLFiJl6fvxdlq4b",
                    }
                  }
                  key=".0"
                  status={Symbol(TODO)}
                  type="HYPOTHESIS"
                >
                  <CardsList
                    filterBy={
                      Object {
                        "needsIndex": 0,
                        "persona": "uOCF2GLFiJl6fvxdlq4b",
                      }
                    }
                    list={Array []}
                    status={Symbol(TODO)}
                    type="HYPOTHESIS"
                  >
                    <div />
                  </CardsList>
                </Connect(CardsList)>
              </div>
            </CardsBoard>
          </Connect(CardsBoard)>
        </li>
        <li
          key="1"
        >
          To write real User Stories, ie have a Customer Driven thinking
          <Connect(CreateCard)
            personaId="uOCF2GLFiJl6fvxdlq4b"
            personaNeedsIndex={1}
          >
            <CreateCard
              onValidate={[Function]}
              personaId="uOCF2GLFiJl6fvxdlq4b"
              personaNeedsIndex={1}
            >
              <div
                className="create-card"
              >
                <button
                  onClick={[Function]}
                >
                  Create new task
                </button>
              </div>
            </CreateCard>
          </Connect(CreateCard)>
          <Connect(CardsBoard)
            type="HYPOTHESIS"
          >
            <CardsBoard
              getCards={[Function]}
              type="HYPOTHESIS"
            >
              <div>
                <Connect(CardsList)
                  filterBy={
                    Object {
                      "needsIndex": 1,
                      "persona": "uOCF2GLFiJl6fvxdlq4b",
                    }
                  }
                  key=".0"
                  status={Symbol(TODO)}
                  type="HYPOTHESIS"
                >
                  <CardsList
                    filterBy={
                      Object {
                        "needsIndex": 1,
                        "persona": "uOCF2GLFiJl6fvxdlq4b",
                      }
                    }
                    list={Array []}
                    status={Symbol(TODO)}
                    type="HYPOTHESIS"
                  >
                    <div />
                  </CardsList>
                </Connect(CardsList)>
              </div>
            </CardsBoard>
          </Connect(CardsBoard)>
        </li>
        <li
          key="2"
        >
          To share at the right time and with the rigth level of information the User Stories with his team
          <Connect(CreateCard)
            personaId="uOCF2GLFiJl6fvxdlq4b"
            personaNeedsIndex={2}
          >
            <CreateCard
              onValidate={[Function]}
              personaId="uOCF2GLFiJl6fvxdlq4b"
              personaNeedsIndex={2}
            >
              <div
                className="create-card"
              >
                <button
                  onClick={[Function]}
                >
                  Create new task
                </button>
              </div>
            </CreateCard>
          </Connect(CreateCard)>
          <Connect(CardsBoard)
            type="HYPOTHESIS"
          >
            <CardsBoard
              getCards={[Function]}
              type="HYPOTHESIS"
            >
              <div>
                <Connect(CardsList)
                  filterBy={
                    Object {
                      "needsIndex": 2,
                      "persona": "uOCF2GLFiJl6fvxdlq4b",
                    }
                  }
                  key=".0"
                  status={Symbol(TODO)}
                  type="HYPOTHESIS"
                >
                  <CardsList
                    filterBy={
                      Object {
                        "needsIndex": 2,
                        "persona": "uOCF2GLFiJl6fvxdlq4b",
                      }
                    }
                    list={Array []}
                    status={Symbol(TODO)}
                    type="HYPOTHESIS"
                  >
                    <div />
                  </CardsList>
                </Connect(CardsList)>
              </div>
            </CardsBoard>
          </Connect(CardsBoard)>
        </li>
        <li
          key="3"
        >
          To have a clear vision on the job he'll have to do AND the job already or currently done
          <Connect(CreateCard)
            personaId="uOCF2GLFiJl6fvxdlq4b"
            personaNeedsIndex={3}
          >
            <CreateCard
              onValidate={[Function]}
              personaId="uOCF2GLFiJl6fvxdlq4b"
              personaNeedsIndex={3}
            >
              <div
                className="create-card"
              >
                <button
                  onClick={[Function]}
                >
                  Create new task
                </button>
              </div>
            </CreateCard>
          </Connect(CreateCard)>
          <Connect(CardsBoard)
            type="HYPOTHESIS"
          >
            <CardsBoard
              getCards={[Function]}
              type="HYPOTHESIS"
            >
              <div>
                <Connect(CardsList)
                  filterBy={
                    Object {
                      "needsIndex": 3,
                      "persona": "uOCF2GLFiJl6fvxdlq4b",
                    }
                  }
                  key=".0"
                  status={Symbol(TODO)}
                  type="HYPOTHESIS"
                >
                  <CardsList
                    filterBy={
                      Object {
                        "needsIndex": 3,
                        "persona": "uOCF2GLFiJl6fvxdlq4b",
                      }
                    }
                    list={Array []}
                    status={Symbol(TODO)}
                    type="HYPOTHESIS"
                  >
                    <div />
                  </CardsList>
                </Connect(CardsList)>
              </div>
            </CardsBoard>
          </Connect(CardsBoard)>
        </li>
      </ul>
    </div>
  </Persona>
</Provider>
`;

exports[`Componant tests Display a persona without children 1`] = `
<div>
  Paul Deligny
  <ul>
    <li
      key="0"
    >
      To have a clear understanding about the product vision & objectives
    </li>
    <li
      key="1"
    >
      To write real User Stories, ie have a Customer Driven thinking
    </li>
    <li
      key="2"
    >
      To share at the right time and with the rigth level of information the User Stories with his team
    </li>
    <li
      key="3"
    >
      To have a clear vision on the job he'll have to do AND the job already or currently done
    </li>
  </ul>
</div>
`;
